lib/ES/BaseRepo.pm

```diff
@@ -0,0 +1,132 @@
+package ES::BaseRepo;
+
+use strict;
+use warnings;
+use v5.10;
+
+use Path::Class();
+use URI();
+use Encode qw(decode_utf8);
+use ES::Util qw(run);
+
+#===================================
+sub new {
+#===================================
+    my ( $class, %args ) = @_;
+
+    my $name = $args{name} or die "No <name> specified";
+    my $url  = $args{url}  or die "No <url> specified for repo <$name>";
+    if ( my $user = $args{user} ) {
+        $url = URI->new($url);
+        $url->userinfo($user);
+    }
+
+    my $dir = $args{dir} or die "No <dir> specified for repo <$name>";
+
+    my $reference_dir = 0;
+    if ($args{reference}) {
+        my $reference_subdir = $url;
+        $reference_subdir =~ s|/$||;
+        $reference_subdir =~ s|:*/*\.git$||;
+        $reference_subdir =~ s/.*[\/:]//g;
+        $reference_dir = $args{reference}->subdir("$reference_subdir.git");
+    }
+
+    bless {
```

Comment:
To be honest I would use explicit `return` instead of relying on implicit one

```perl
my $self = bless {
...

return $self;
```

Reply:
:+1: